CUDADataFormats/Common/interface/HeterogeneousSoA.h

```diff
@@ -0,0 +1,247 @@
+#ifndef CUDADataFormatsCommonHeterogeneousSoA_H
+#define CUDADataFormatsCommonHeterogeneousSoA_H
+
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
+
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "HeterogeneousCore/CUDAServices/interface/CUDAService.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/copyAsync.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+// a heterogeneous unique pointer...
+template<typename T>
+class HeterogeneousSoA {
+public:
+
+  using Product = T;
+
+  HeterogeneousSoA() = default; // make root happy
+  ~HeterogeneousSoA() = default;
+  HeterogeneousSoA(HeterogeneousSoA&&) = default;
+  HeterogeneousSoA& operator=(HeterogeneousSoA&&) = default;
+
+  explicit HeterogeneousSoA(cudautils::device::unique_ptr<T> && p) : dm_ptr(std::move(p)) {}
+  explicit HeterogeneousSoA(cudautils::host::unique_ptr<T> && p) : hm_ptr(std::move(p)) {}
+  explicit HeterogeneousSoA(std::unique_ptr<T> && p) : std_ptr(std::move(p)) {}
+
+
+  auto const * get() const {
+    return dm_ptr ? dm_ptr.get() : (hm_ptr ? hm_ptr.get() : std_ptr.get());
+  }
+
+  auto const & operator*() const {
+    return *get();
+  }
+
+  auto const * operator->() const {
+    return get();
+  }
+
+
+  auto * get() {
+    return dm_ptr ? dm_ptr.get() : (hm_ptr ? hm_ptr.get() : std_ptr.get());
+  }
+
+  auto &  operator*() {
+    return *get();
+  }
+
+  auto * operator->() {
+    return get();
+  }
+
+
+
+  // in reality valid only for GPU version...
+  cudautils::host::unique_ptr<T>
+  toHostAsync(cuda::stream_t<>& stream) const {
+    assert(dm_ptr);
+    edm::Service<CUDAService> cs;
+    auto ret = cs->make_host_unique<T>(stream);
+    cudaCheck(cudaMemcpyAsync(ret.get(), dm_ptr.get(), sizeof(T), cudaMemcpyDefault, stream.id()));
+    return ret;
+  }
+
+
+
+private:
+  // a union wan't do it, a variant will not be more efficienct
+  cudautils::device::unique_ptr<T> dm_ptr; //!
+  cudautils::host::unique_ptr<T> hm_ptr; //!
+  std::unique_ptr<T> std_ptr; //!
```

Comment:
I have mixed feelings about this approach.

On the one hand, it is not very flexible to extend if we want to support different types of accelerators (are we going to have a pair of pointers for each possible case ?) and somewhat redundant (we should find a better way that having both an `std::unique_ptr<T>` and an `cudautils::host::unique_ptr<T>`).

On the other hand, it could be used to make the host<-->device copy "in place" as far as the framework is concerned: a single framework product could own the memory on the host and (multiple) devices, and copy it on demand.

Reply:
the alternative is shared_ptr approach:
one pointer to T, 
one pointer to a deleter

in any case we have to move away from default std::unique_ptr as it uses ```new```&```delete``` that calls ```T```'s constructor&destructor while in all other cases one uses ```malloc```&```free``` (when needed initialization is explicit and run on target)